generate/seccomp/syscall_compare.go

```diff
@@ -0,0 +1,143 @@
+/*******************************************************************************
+This file contains functions for determining the action to be taken for adding
+				syscalls rules to the seccomp configuration.
+*******************************************************************************/
+
+package seccomp
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	types "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func decideCourseOfAction(newSyscall *types.Syscall, syscalls []types.Syscall) (string, error) {
+	ruleForSyscallAlreadyExists := false
+
+	var sliceOfDeterminedActions []string
+	for i, syscall := range syscalls {
+		if syscall.Name == newSyscall.Name {
+			ruleForSyscallAlreadyExists = true
+
+			if identical(newSyscall, &syscall) {
+				sliceOfDeterminedActions = append(sliceOfDeterminedActions, nothing)
+			}
+
+			if sameAction(newSyscall, &syscall) {
+				if bothHaveArgs(newSyscall, &syscall) {
+					sliceOfDeterminedActions = append(sliceOfDeterminedActions, seccompAppend)
+				}
+				if onlyOneHasArgs(newSyscall, &syscall) {
+					if firstParamOnlyHasArgs(newSyscall, &syscall) {
+						sliceOfDeterminedActions = append(sliceOfDeterminedActions, "overwrite:"+strconv.Itoa(i))
+					} else {
+						sliceOfDeterminedActions = append(sliceOfDeterminedActions, nothing)
+					}
+				}
+			}
+
+			if !sameAction(newSyscall, &syscall) {
+				if bothHaveArgs(newSyscall, &syscall) {
+					if sameArgs(newSyscall, &syscall) {
+						sliceOfDeterminedActions = append(sliceOfDeterminedActions, "overwrite:"+strconv.Itoa(i))
+					}
+					if !sameArgs(newSyscall, &syscall) {
+						sliceOfDeterminedActions = append(sliceOfDeterminedActions, seccompAppend)
+					}
+				}
+				if onlyOneHasArgs(newSyscall, &syscall) {
+					sliceOfDeterminedActions = append(sliceOfDeterminedActions, seccompAppend)
+				}
+				if neitherHasArgs(newSyscall, &syscall) {
+					sliceOfDeterminedActions = append(sliceOfDeterminedActions, "overwrite:"+strconv.Itoa(i))
+				}
+			}
+		}
+	}
+
+	if !ruleForSyscallAlreadyExists {
+		sliceOfDeterminedActions = append(sliceOfDeterminedActions, seccompAppend)
+	}
+
+	// Nothing has highest priority
+	for _, determinedAction := range sliceOfDeterminedActions {
+		if determinedAction == nothing {
+			return determinedAction, nil
+		}
+	}
+
+	// Overwrite has second highest priority
+	for _, determinedAction := range sliceOfDeterminedActions {
+		if strings.Contains(determinedAction, seccompOverwrite) {
+			return determinedAction, nil
+		}
+	}
+
+	// Append has the lowest priority
+	for _, determinedAction := range sliceOfDeterminedActions {
+		if determinedAction == seccompAppend {
+			return determinedAction, nil
+		}
+	}
+
+	return "error", fmt.Errorf("Trouble determining action: %s", sliceOfDeterminedActions)
```

Comment:
On Thu, Aug 04, 2016 at 02:42:49PM -0700, Grant Seltzer Richman wrote:
> > +	return "error", fmt.Errorf("Trouble determining action: %s", sliceOfDeterminedActions)
> 
> Yea, i'm not sure what the idiomatic way of doing this was. It
> didn't feel right to return `determinedAction`, it's certainly
> another check incase the user isn't properly checking errors. Would
> you use just an empty string?

I see ‘"", err’ frequently, although I don't read a lot of Go.  In
master:

  $ git grep 'return .*""'
  Godeps/_workspace/src/github.com/Sirupsen/logrus/entry.go:              return "", err
  Godeps/_workspace/src/github.com/Sirupsen/logrus/hooks/sentry/sentry.go:                return "", false
  Godeps/_workspace/src/github.com/Sirupsen/logrus/hooks/sentry/sentry.go:                return "", false
  Godeps/_workspace/src/github.com/blang/semver/range.go:         return nil, fmt.Errorf("Could not parse comparator %q in %q", opStr, strings.Join([]string{opStr, vStr}, ""))
  Godeps/_workspace/src/github.com/blang/semver/range.go:         return nil, fmt.Errorf("Could not parse version %q in %q: %s", vStr, strings.Join([]string{opStr, vStr}, ""), err)
  Godeps/_workspace/src/github.com/blang/semver/range.go:         return "", "", fmt.Errorf("Could not get version from string: %q", s)
  Godeps/_workspace/src/github.com/blang/semver/semver.go:                return "", errors.New("Buildversion is empty")
  Godeps/_workspace/src/github.com/blang/semver/semver.go:                return "", fmt.Errorf("Invalid character(s) found in build meta data %q", s)
  Godeps/_workspace/src/github.com/blang/semver/semver_test.go:   return PRVersion{"", i, true}
  Godeps/_workspace/src/github.com/urfave/cli/context.go: return ""
  Godeps/_workspace/src/github.com/urfave/cli/context.go: return ""
  Godeps/_workspace/src/github.com/urfave/cli/context.go: return ""
  Godeps/_workspace/src/github.com/urfave/cli/flag.go:    return "", usage